lib/workers/repository/update/pr/changelog/source-bitbucket.ts

```diff
@@ -0,0 +1,159 @@
+  // istanbul ignore if
+  if (cachedResult !== undefined) {
+    return cachedResult;
+  }
+  const promisedRes = getTags(repository);
+  memCache.set(cacheKey, promisedRes);
+  return promisedRes;
+}
+
+export async function getChangeLogJSON(
+  config: BranchUpgradeConfig
+): Promise<ChangeLogResult | null> {
+  const versioning = config.versioning!;
+  const currentVersion = config.currentVersion!;
+  const newVersion = config.newVersion!;
+  const sourceUrl = config.sourceUrl!;
+  const packageName = config.packageName!;
+  const sourceDirectory = config.sourceDirectory!;
+
+  logger.trace('getChangeLogJSON for bitbucket');
+  const version = allVersioning.get(versioning);
+
+  const parsedUrl = URL.parse(sourceUrl);
+  const protocol = parsedUrl.protocol!;
+  const host = parsedUrl.host!;
+  const pathname = parsedUrl.pathname!;
+
+  logger.trace({ protocol, host, pathname }, 'Protocol, host, pathname');
+  const baseUrl = protocol.concat('//', host, '/');
+  const apiBaseUrl = protocol.concat('//', 'api.', host, '/');
+  const repository = pathname
+    .slice(1)
+    .replace(regEx(/\/$/), '')
+    .replace(regEx(/\.git$/), '');
+
+  if (repository.split('/').length < 2) {
+    logger.info({ sourceUrl }, 'Invalid bitbucket URL found');
+    return null;
+  }
+
+  const releases = config.releases ?? (await getInRangeReleases(config));
+  if (!releases?.length) {
+    logger.debug('No releases');
+    return null;
+  }
+
+  // This extra filter/sort should not be necessary, but better safe than sorry
+  const validReleases = [...releases]
+    .filter((release) => version.isVersion(release.version))
+    .sort((a, b) => version.sortVersions(a.version, b.version));
+
+  if (validReleases.length < 2) {
+    logger.debug('Not enough valid releases');
+    return null;
+  }
+
+  let tags: string[];
+
+  async function getRef(release: Release): Promise<string | null> {
+    if (!tags) {
+      tags = await getCachedTags(apiBaseUrl, versioning, repository);
+    }
+    const regex = regEx(`(?:${packageName}|release)[@-]`, undefined, false);
+    const tagName = tags
+      .filter((tag) => version.isVersion(tag.replace(regex, '')))
+      .find((tag) => version.equals(tag.replace(regex, ''), release.version));
+    if (tagName) {
+      return tagName;
+    }
+    if (release.gitRef) {
+      return release.gitRef;
+    }
+    return null;
+  }
+
+  function getCacheKey(prev: string, next: string): string {
+    return `${slugifyUrl(sourceUrl)}:${packageName}:${prev}:${next}`;
+  }
+
+  const changelogReleases: ChangeLogRelease[] = [];
+  // compare versions
+  const include = (v: string): boolean =>
+    version.isGreaterThan(v, currentVersion) &&
+    !version.isGreaterThan(v, newVersion);
+
+  for (let i = 1; i < validReleases.length; i += 1) {
+    const prev = validReleases[i - 1];
+    const next = validReleases[i];
+    if (include(next.version)) {
+      let release = await packageCache.get(
```

Comment:
Could be easier readable if you extract this into a function and then cache using annotations

Reply:
agree, but my thoughts were to keep it as close to the github and gitlab impl so that the larger refactoring effort was consistent

Reply:
no, we should make this as reusable as possible now and adapt existing implementations later